es6/animations.js

```diff
@@ -5,180 +5,309 @@ import {
     checkIsValidAnimation,
     checkIsValidSequence,
     EasingFunctions,
+    immutify,
+    computeAnimatedState,
+    mapObj,
+    findMissingKey,
 } from './util.js'
 
+import isEqual from 'lodash.isequal'
 
-const unit_tick = ({start_time, end_time, duration, start_state, end_state, amt, curve='linear', unit=null}, key) => {
-    var {start_state, amt, end_state} = checked_animation_amt({start_state, end_state, amt, key})
-    var {start_time, end_time, duration} = checked_animation_duration({start_time, end_time, duration})
+const fmtWithUnit = (val, unit) => {
+    return unit ? `${val}${unit}` : val
+}
+
+const tick_func = ({duration, start_state, delta_state, end_state,
+                    curve='linear', unit=null}) => {
 
     const curve_func = EasingFunctions[curve]
 
-    return (delta) => {
-        let new_state
-        // These over-boundary cases happen because we need to render
-        // an extra frame before/after start/end times to finish off
-        // animations who's (durations) % (frame rate) != 0.
-        if (delta < 0) {
-            new_state = start_state
-        } else if (delta >= duration) {
-            new_state = end_state
-        } else {
-            // tick progression function, the core math at the heart of animation
-            new_state = start_state + curve_func(delta/duration)*amt
+    return (time_elapsed) => {
+        if (time_elapsed < 0) {
+            return start_state
         }
-        return unit ? `${new_state}${unit}` : new_state
+        if (time_elapsed > duration) {
+            return end_state
+        }
+        const curve_value = curve_func(time_elapsed/duration)
+        const new_state = start_state + curve_value * delta_state
+
+        return fmtWithUnit(new_state, unit)
     }
 }
 
-const checked_animation_duration = ({start_time, duration, end_time}) => {
-    if ([start_time, end_time, duration].filter(a => typeof(a) == 'number').length < 2) {
-        console.log({start_time, end_time, duration})
-        throw 'Need at least 2/3 to calculate animation: start_time, end_time, duration'
-    }
+export const Become = ({path, state, start_time, end_time, duration}) => {
+    if (start_time === undefined) start_time = Date.now()
 
-    if (start_time === undefined)
-        start_time = end_time - duration
-    if (end_time === undefined)
-        end_time = start_time + duration
-    if (duration === undefined)
-        duration = end_time - start_time
+    if (end_time === undefined && duration === undefined) {
+        duration = Infinity
+        end_time = Infinity
+    }
 
-    if (start_time + duration != end_time) {
-        console.log({start_time, end_time, duration})
-        throw 'Conflicting values, Animation end_time != start_time + duration'
+    if (end_time !== Infinity || duration !== Infinity) {
+        if (exactlyOneIsUndefined(duration, end_time)) {
```

Comment:
Wouldn't be better if we split the `computeTheOther` in `computeEndtime` and `computeDuration` and change `exactlyOneIsUndefined` to `isUdefined`, somethig like:
```javascript
if (isUdefined(duration)) {
    duration = computeDuration(start_time, end_time)
} else if (isUdefined(end_time)) {
    end_time = computeEndtime(start_time, duration)
} else {
    // console.log({path, state, start_time, end_time, duration})
    throw 'Invalid call to Become: you may define end_time or duration, but not both.'
}
```
Because in `computeTheOther` we ask again for which one is undefined in both cases, for objects and numbers.

Reply:
I see that the computeOther works for states instead for just time.

Reply:
I still think something like this is the cleanest encapsulation, it's clear that only these values are being checked, and the logic is all contained in two functions with nice, specific error messages.

```python
{start_time, end_time, duration} = checkedDeltaTime({start_time, end_time, duration})
{start_state, end_state, delta} = checkedDeltaState({start_state, end_state, delta})
...
```

Reply:
This is the case for the time args, but the state checks are different for every case, so that can't really be abstracted into a function.